### ISSUE TITLE
Allow users to specify how frequent PRs can be created

### ISSUE THREAD

**Issue by fthomas** [fthomas/scala-steward#1316]
Scala Steward currently creates pull requests as soon as it detects that a new version is available. This leads to a lot of PRs for projects that have many dependencies or dependecies that are very frequently updated. To reduce the noise this creates for maintainers of such projects, we should add the ability control how frequent Scala Steward is allowed to create PRs in a repo via an option in `.scala-steward.conf`. This option could be a simple enum with the cases `asap | daily | weekly | monthly`.

See https://twitter.com/olafurpg/status/1227392299904258048 for more context.

**Comment by dwijnand** [comment]
As Renato mentions (https://twitter.com/renatocaval/status/1227541373064286208) maybe it could bundle all the updates in a single PR too.  Of course, that also trades off with testing the individual upgrades in isolation.  But you also save on CI consumption.  Trade-offs, trade-offs...

**Comment by kiranbayram** [comment]
I'm interested in knowing if/where we want to persist the state if a successful PR creation has completed for that time period (for daily/weekly/monthly)

**Comment by renatocaval** [comment]
Here are my thoughts about it ...

If we have a longer period, we won't reduce the number of notifications nor the pressure on CI. We will only post-pone it to a later date. Then, that day will be the day to watch CI finish n builds and merge n PRs, etc.

We will reduce the noise when a library is updated three times in the week. We will only get the last update. But that's not the most common case. 

That's why I had this idea of accumulating everything in one PR. This has the drawback that different updates may break the build, but if we have one commit per update, we can run `git bisect` locally to try to find the offending library. That will be painful, but it's part of the trade-off. 

Anyway, cool that we are discussing this. I think there is some room for improvements and we are moving in a good direction.

**Comment by fthomas** [comment]
I understand the arguments for bundling updates into one PR but I'm not a fan of that. I summarised my arguments against that feature in https://github.com/fthomas/scala-steward/issues/413#issuecomment-585410159. They basically boil down to "this makes Scala Steward more complicated". But that doesn't mean this feature won't be added to Scala Steward, only that it is probably not me who adds it. :-)

However, I think we can decouple bundling updates and giving users control over the frequency of PR creation. I'm confident that the latter doesn't require a ton of changes and that it can be implemented so that it is independent of a future "bundle updates" feature.

I'd also like to mention that this feature without bundling updates might not reduce the overall number of notifications, but it does reduce the number of days when these notifications are created. I could imagine that Scala Steward is perceived less noisy when it only creates PRs once a week instead of everyday.

@kiranbayram I would store the date a PR was created in `PullRequestRepository` and then use the latest date there to decide if we are allowed to create new PRs.

**Comment by kiranbayram** [comment]
I have two questions:
 - Would you use `java.time.LocalDateTime` in app context to compare dates?
 - How would you serialize the date? Some form of formatted string (`"2007-12-03T10:15:30"`) or as an `Instant` (epoch seconds)?

**Comment by fthomas** [comment]
I've been using `DateTimeAlg.currentTimeMillis` before to get the current milliseconds since epoch and serialized it as `Long`. For comparison I've been using `FiniteDuration`.

My current thinking about how this could implemented is that we could remove all `DependencyOutdated` states in [`PruningAlg.checkUpdateStates`](https://github.com/fthomas/scala-steward/blob/12236e31327666180d13e34039ca2eb74c043a54/modules/core/src/main/scala/org/scalasteward/core/update/PruningAlg.scala#L126-L143) if the duration between the creation of the last PR and now is smaller than the duration specified in the `.scala-steward.conf`. And that's it. No new PRs will be created but existing ones would still be updated in case of merge conflicts.

**Comment by mzuehlke** [comment]
One question that come to my mind. Do we want to have this time period per repo or per dependency ?
They have a different effect:
* **per repo**: all update are coming on the same day once per week / once per month. All other days are without "noise"
* **per dependency**: the same rhythm of updates as now. Only dependencies that are changing ver often are "throttled"

**Comment by fthomas** [comment]
That's an interesting question! I've assumed **per repo** until now but **per dependency** sounds interesting too. One downside I see with **per dependency** is that if for example a library publishes a botched version 1.0 and shortly after that a fixed version 1.1, **per repo** will likely create a PR for version 1.1 only while **per dependency** will create a PR for 1.0 and only after the time period one for 1.1.

**Comment by olafurpg** [comment]
@renatocaval bundling upgrades into one PR sounds like a separate feature request to me. It's not something I'm personally interested in since if the CI fails you would need to manually track down which upgrade caused the failure.

@fthomas thank you for opening this ticket! I want to register my interest in using this feature ✋ My biggest gripe is getting PRs for version numbers that are short-lived because the library publishes frequently (many are my own libraries 😅).

Just one example, over the past week I have published four versions of olafurpg/metaconfig in roughly 1-2 intervals while working on a new feature in the scalameta/metals repository. Metaconfig is also used in my other projects (scalameta/scalafmt, scalacenter/scalafix, ...) so I also received four upgrade PRs in those repositories even if they don't need the new functionality in Metaconfig (at least not yet). If Scala Steward upgrade PRs only came ~monthly, then I would most likely only receive one upgrade PR for Metaconfig.

By running the upgrade PRs less frequently, we may be able to at least skip a few redundant PRs for upgrades to short-lived versions.

Just brainstorming here, it might be nice to have an option to say something like

* only send an upgrade PR when the release is older than X number of days (I'm not sure if there is reliable data to implement this), with a maximum limit of Y days (so that we don't wait indefinitely to upgrade)
* only send upgrade PRs on specific weekdays. For example, I would prefer not to get the upgrade PRs over the weekend.

**Comment by olafurpg** [comment]
To give one more example from yesterday, I got three Scala Steward PRs to upgrade the same library

* https://github.com/scalameta/mdoc/pull/295 MUnit v0.5.0
* https://github.com/scalameta/mdoc/pull/296 MUnit v0.5.1
* https://github.com/scalameta/mdoc/pull/297 MUnit v0.5.2

The v0.5.1 release failed for Scala.js 0.6.x due to build mis-configuration but succeeded for all other modules. I use MUnit in four other Scalameta repositories so I ended up getting quite  a lot of PRs that are redundant.

**Comment by fthomas** [comment]
> only send an upgrade PR when the release is older than X number of days (I'm not sure if there is reliable data to implement this), with a maximum limit of Y days (so that we don't wait indefinitely to upgrade)

That's an interesting suggestion to reduce the overall noise. We don't have the data yet to implement this but we could certainly record when Scala Steward sees  a new version for the first time. But I also think that this is a separate feature that could be implemented independently.

> only send upgrade PRs on specific weekdays. For example, I would prefer not to get the upgrade PRs over the weekend.

I like that too. Allowing cron expressions instead of or in addition to the `asap | daily | weekly | monthly` enum would make this possible.

**Comment by fthomas** [comment]
FYI: I started working on this feature in #1336.
